shared/modules/transaction.utils.js

```diff
@@ -61,3 +81,53 @@ export function txParamsAreDappSuggested(transaction) {
       transaction?.dappSuggestedGasFees?.maxFeePerGas === maxFeePerGas)
   );
 }
+
+/**
+ * Determines the type of the transaction by analyzing the txParams.
+ * This method will return one of the types defined in shared/constants/transactions
+ * It will never return TRANSACTION_TYPE_CANCEL or TRANSACTION_TYPE_RETRY as these
+ * represent specific events that we control from the extension and are added manually
+ * at transaction creation.
+ *
+ * @param {Object} txParams - Parameters for the transaction
+ * @param {EthQuery} query - EthQuery instance
+ * @returns {InferTransactionTypeResult}
+ */
+export async function determineTransactionType(txParams, query) {
+  const { data, to } = txParams;
+  let name;
+  try {
+    name = data && hstInterface.parseTransaction({ data }).name;
+  } catch (error) {
+    log.debug('Failed to parse transaction data.', error, data);
+  }
+
+  const tokenMethodName = [
+    TRANSACTION_TYPES.TOKEN_METHOD_APPROVE,
+    TRANSACTION_TYPES.TOKEN_METHOD_TRANSFER,
+    TRANSACTION_TYPES.TOKEN_METHOD_TRANSFER_FROM,
```

Comment:
also not something we need to do in this PR but I just created this [issue](https://github.com/MetaMask/metamask-extension/issues/13855) and we should consider adding `safeTransferFrom` here soon.

Reply:
Yeah, I don't want to make this about anything other than extracting this method. A future PR should be launched to change this behavior.